### PR TITLE
Refactor configure command

### PR DESCRIPTION
**Pull request by andrzejtp**
Follow https://docs.digitalocean.com/reference/doctl/reference/auth/ example.

It also addresses:

https://github.com/boardswarm/boardswarm/issues/28
https://github.com/boardswarm/boardswarm/issues/36

### CODE REVIEW

File: boardswarm-cli/src/main.rs [hunk 1]
```diff
@@ -477,42 +492,33 @@ enum Command {
...
+    });
+    if let Some(s) = config.find_server(&instance) {
+        bail!("Authentication for {} already exists.", s.name);
     }
 
-    // UGGGG
+    let uri = init_args
+        .uri
+        .as_ref()
+        .ok_or_else(|| anyhow!("New authentication needs a uri"))?;
```
Comment (sjoerdsimons) [inline boardswarm-cli/src/main.rs:512]: If you split the argument structs for `init` and `modify` then clap can already enforce this for you (make url not an Option)

File: boardswarm-cli/src/main.rs [hunk 2]
```diff
@@ -477,42 +492,33 @@ enum Command {
...
-        None => config.default_server_mut(),
-    };
-
-    match (current_server.is_some(), opts.new) {
-        (true, true) => bail!("Instance already exists"),
-        (false, false) => bail!("Instance not found, use --new to create a new one"),
-        _ => (),
+    if let (Some(_), Some(_)) = (&init_args.token, &init_args.token_file) {
+        bail!("Cannot use --token and --token_file simultaneously");
     }
```
Comment (sjoerdsimons) [inline boardswarm-cli/src/main.rs:517]: See other note about using clap's conflicts_with

**Review by sjoerdsimons — CHANGES_REQUESTED** [review]

File: boardswarm-cli/src/main.rs [hunk 3]
```diff
@@ -549,38 +552,134 @@ async fn run_configure(
...
-        if opts.default {
-            config.set_default(&instance);
+    };
+    if config.find_server(&instance).is_none() {
+        bail!("Authentication for {} does not exist", instance);
+    }
+
+    if let Some(s) = config.default_server() {
+        if s.name == instance && !force {
+            bail!("Trying to remove default authentication. Use --force to proceed");
```
[the remark this answers is not in the document]
Comment (andrzejtp) [reply boardswarm-cli/src/main.rs:631]: Trying to not let the user shoot themselves in the foot

File: boardswarm-cli/src/main.rs [hunk 4]
```diff
@@ -549,38 +552,134 @@ async fn run_configure(
...
+    config.write().await?;
+
+    Ok(())
+}
+
+async fn run_list(config: config::Config) -> anyhow::Result<()> {
+    let default = config.default_server();
+
+    for s in config.servers() {
+        if let Some(default) = default {
```
[the remark this answers is not in the document]
Comment (andrzejtp) [reply boardswarm-cli/src/main.rs:584]: to print an asterisk before the default server's name. Can I rely on the default _always_ being the first in the list?

File: boardswarm-cli/src/main.rs [hunk 5]
```diff
@@ -477,42 +492,33 @@ enum Command {
...
     instance: Option<String>,
-    opts: &ConfigureArg,
+    init_args: AuthArg,
 ) -> anyhow::Result<()> {
-    if opts.new && instance.is_none() {
-        bail!("New instances require an instance argument");
+    let mut is_default = false;
+    let instance = instance.unwrap_or_else(|| {
+        is_default = true;
+        "default".to_string()
```
[the remark this answers is not in the document]
Comment (andrzejtp) [reply boardswarm-cli/src/main.rs:503]: This is how I understand doctl works. We don't need to follow it exactly, though. Might just as well require -i for "init" subcommand

File: boardswarm-cli/src/main.rs [hunk 6]
```diff
@@ -477,42 +492,33 @@ enum Command {
...
+    });
+    if let Some(s) = config.find_server(&instance) {
+        bail!("Authentication for {} already exists.", s.name);
```
[the remark this answers is not in the document]
Comment (andrzejtp) [reply boardswarm-cli/src/main.rs:506]: good point!

File: boardswarm-cli/src/main.rs [hunk 7]
```diff
@@ -637,14 +736,30 @@ async fn main() -> anyhow::Result<()> {
...
-        Command::Configure(ref configure) => {
+        Command::Auth { command } => {
             let config = config
                 .clone()
                 .unwrap_or_else(|| config::Config::new(config_path));
 
-            return run_configure(config, opt.instance, configure).await;
+            match command {
+                AuthCommand::Init(init_args) => {
+                    return run_init(config, opt.instance, init_args).await;
```
[the remark this answers is not in the document]
Comment (andrzejtp) [reply boardswarm-cli/src/main.rs:746]: sure np

File: boardswarm-cli/src/main.rs [hunk 8]
```diff
@@ -549,38 +552,134 @@ async fn run_configure(
...
+    let instance = match instance {
+        Some(s) => s,
+        None => {
+            bail!("-i/--instance required to specify the new default");
         }
+    };
+
+    if let (Some(_), Some(_)) = (&modify_args.token, &modify_args.token_file) {
+        bail!("Cannot use --token and --token_file simultaneously");
     }
```
[the remark this answers is not in the document]
Comment (andrzejtp) [reply boardswarm-cli/src/main.rs:655]: I guess relying on clap is better

File: boardswarm-cli/src/main.rs (hunk 2 above)
Comment (andrzejtp) [reply boardswarm-cli/src/main.rs:517]: yup

File: boardswarm-cli/src/main.rs (hunk 1 above)
Comment (andrzejtp) [reply boardswarm-cli/src/main.rs:512]: ok

File: boardswarm-cli/src/main.rs [hunk 9]
```diff
@@ -196,6 +196,37 @@ async fn rock_download_boot(volume: &mut DeviceVolume, path: &Path) -> anyhow::R
...
+#[derive(Debug, clap::Parser)]
+struct AuthArg {
+    /// Configure the url
+    #[clap(short, long)]
+    uri: Option<Uri>,
+    /// Configure the JWT auth token
+    #[clap(short, long)]
+    token: Option<String>,
+    /// Read new JWT auth token from the given file
+    #[clap(long)]
```
[the remark this answers is not in the document]
Comment (andrzejtp) [reply boardswarm-cli/src/main.rs:208]: ok

File: boardswarm-cli/src/main.rs [hunk 10]
```diff
@@ -549,38 +552,134 @@ async fn run_configure(
...
+async fn run_modify(
+    mut config: config::Config,
+    instance: Option<String>,
+    modify_args: AuthArg,
+) -> anyhow::Result<()> {
+    let instance = match instance {
+        Some(s) => s,
+        None => {
+            bail!("-i/--instance required to specify the new default");
         }
```
[the remark this answers is not in the document]
Comment (sjoerdsimons) [reply boardswarm-cli/src/main.rs]: looks like you worked it out :) sorry that was meant to be `a instance`

File: boardswarm-cli/src/main.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sjoerdsimons) [reply boardswarm-cli/src/main.rs]: RIght that's an implementation detail in the client crate indeed; but yeah you can rely on it. Probably worth documenting that behaviour on the .servers() function

File: boardswarm-cli/src/main.rs [hunk 11]
```diff
@@ -559,38 +560,97 @@ async fn run_configure(
...
+    Ok(())
+}
+
+async fn run_auth_list(config: config::Config, instance: Option<String>) -> anyhow::Result<()> {
+    let servers = config.servers().filter(|s| {
+        if let Some(name) = &instance {
+            s.name == name.as_str()
+        } else {
+            true
         }
```
[the remark this answers is not in the document]
Comment (andrzejtp) [reply boardswarm-cli/src/main.rs:590]: +1 for shortening. As for filtering I think it does make sense. Suppose you have two screenfuls of servers and you want to be reminded what the uri of your favourite one is. You _can_ grep, but why not request listing only the right one?

**Review by sjoerdsimons — APPROVED** [review]

**Pull request merged**